src/ripple/basics/SlabAllocator.h

```diff
@@ -100,91 +126,109 @@ class SlabAllocator
+        @param ptr The pointer to the chunk of memory being deallocated.
+    */
+    [[nodiscard]] bool
+    deallocate(SlabBlock* slab, std::uint8_t* ptr) noexcept
+    {
+        assert(ptr != nullptr);
 
-            @note This is a dangerous, private interface; the item being
-                  returned should belong to this allocator. Debug builds
-                  will check and assert if this is not the case. Release
-                  builds will not.
-         */
-        void
-        deallocate(std::uint8_t* ptr) noexcept
         {
-            assert(own(ptr));
+            std::lock_guard l(slab->m_);
 
-            std::lock_guard l(m_);
+            if (auto const d = slab->data_.load();
+                (ptr < d) || (ptr >= d + slabSize_))
+                return false;
 
-            // Use memcpy to avoid unaligned UB
-            // (will optimize to equivalent code)
-            std::memcpy(ptr, &l_, sizeof(std::uint8_t*));
-            l_ = ptr;
-        }
-    };
+            // It looks like this objects belongs to us. Ensure that it
+            // doesn't extend past the end of our buffer:
+            assert(ptr + itemSize_ <= slab->data_ + slabSize_);
 
-private:
-    // A linked list of slabs
-    std::atomic<SlabBlock*> slabs_ = nullptr;
+            // We use memcpy to avoid UB for unaligned accesses
+            std::memcpy(ptr, &slab->l_, sizeof(std::uint8_t*));
+            slab->l_ = ptr;
 
-    // The alignment requirements of the item we're allocating:
-    std::size_t const itemAlignment_;
+            if (++slab->avail_ != itemCount_)
+                return true;
 
-    // The size of an item, including the extra bytes requested and
-    // any padding needed for alignment purposes:
-    std::size_t const itemSize_;
+            // The slab is now fully unused. We're going to release its buffer:
+            ptr = slab->data_.exchange(nullptr);
+            slab->avail_ = 0;
+        }
 
-    // The size of each individual slab:
-    std::size_t const slabSize_;
+        // Release the memory of this fully unused block back to the system
+        // without holding a lock.
+        boost::alignment::aligned_free(ptr);
+        return true;
+    }
 
 public:
     /** Constructs a slab allocator able to allocate objects of a fixed size
 
-        @param count the number of items the slab allocator can allocate; note
-                     that a count of 0 is valid and means that the allocator
-                     is, effectively, disabled. This can be very useful in some
-                     contexts (e.g. when mimimal memory usage is needed) and
-                     allows for graceful failure.
+        @param extra The number of extra bytes per item, on top of sizeof(Type)
+        @param alloc The number of bytes to allocate for this slab
+        @param align The alignment of returned pointers, normally alignof(Type)
      */
     constexpr explicit SlabAllocator(
```

Comment:
I don't think this slab allocator will ever be used in a `constexpr` environment. I might remove the `constexpr` tags. They don't do any harm, but they also don't add value.

Reply:
Once we get fully conforming C++20 compilers, I _believe_ (can't test that I'm right) that the `slabber` variable in SHAMapitem.h, line 117 could be decorated with `constinit`, as long as this constructor remains `constexpr`, which would ensure that this constructor runs at compile-time.